### PR TITLE
SLING-12417- [Compiled Scripts] Add healthcheck for loading capabilities

### PR DESCRIPTION
**Pull request by joerghoh**

### CODE REVIEW

File: src/main/java/org/apache/sling/servlets/resolver/internal/bundle/BundledScriptTracker.java [hunk 1]
```diff
@@ -756,4 +805,16 @@ private Set<BundledRenderUnitCapability> reduce(Set<BundledRenderUnitCapability>
...
         return newSet;
     }
+
+
+    @ObjectClassDefinition
+    public @interface BundledScriptTrackerConfig {
+        
+        @AttributeDefinition(name="Mandatory Bundles", description="for all of the provided symbolic bundle names the "
+                + "registration process must have been completed successfully for the healthcheck to report ok")
+        String[] mandatoryBundles();
```
Comment (raducotescu) [inline src/main/java/org/apache/sling/servlets/resolver/internal/bundle/BundledScriptTracker.java:815]:
```suggestion
        @AttributeDefinition(name="Mandatory Bundles", description="A list of symbolic bundle names the for which the "
                + "script registration process must have been successfully completed for the health check to report ok.")
        String[] mandatoryBundles();
```

File: src/main/java/org/apache/sling/servlets/resolver/internal/bundle/BundledScriptTracker.java [hunk 2]
```diff
@@ -136,10 +154,23 @@ protected void deactivate() {
...
+            healthCheckRegistration = null;
+        }
         bundleContext.set(null);
         dispatchers.set(null);
     }
 
+    @SuppressWarnings({ "rawtypes", "unchecked" })
+    ServiceRegistration<HealthCheck> registerHealthCheck(String[] tags) {
+        Dictionary props = new Hashtable();
+        props.put(HealthCheck.NAME, "Sightly BundledScriptTracker Healthcheck");
```
Comment (raducotescu) [inline src/main/java/org/apache/sling/servlets/resolver/internal/bundle/BundledScriptTracker.java:168]: This health check doesn't have anything to do with Sightly/HTL.

File: src/main/java/org/apache/sling/servlets/resolver/internal/bundle/BundledScriptTracker.java [hunk 3]
```diff
@@ -557,6 +589,23 @@ public void removedBundle(Bundle bundle, BundleEvent event, List<ServiceRegistra
...
         regs.forEach(ServiceRegistration::unregister);
         refreshDispatcher(Collections.emptyList());
+        registeredBundles.remove(bundle.getSymbolicName());
+    }
+    
+    @Override
+    public Result execute() {
+        
+        if (expectedBundles == null) {
+            return new Result(Result.Status.OK,"Healthcheck not configured");
```
Comment (raducotescu) [inline src/main/java/org/apache/sling/servlets/resolver/internal/bundle/BundledScriptTracker.java:599]:
```suggestion
            return new Result(Result.Status.OK,"Health check is not configured.");
```

File: src/main/java/org/apache/sling/servlets/resolver/internal/bundle/BundledScriptTracker.java [hunk 4]
```diff
@@ -557,6 +589,23 @@ public void removedBundle(Bundle bundle, BundleEvent event, List<ServiceRegistra
...
+        }
+        
+        if (registeredBundles.containsAll(expectedBundles)) {
+            return new Result(Result.Status.OK,"all expected bundles registered");
```
Comment (raducotescu) [inline src/main/java/org/apache/sling/servlets/resolver/internal/bundle/BundledScriptTracker.java:603]:
```suggestion
            return new Result(Result.Status.OK,"All expected bundles have registered their scripts.");
```

File: src/main/java/org/apache/sling/servlets/resolver/internal/bundle/BundledScriptTracker.java [hunk 5]
```diff
@@ -557,6 +589,23 @@ public void removedBundle(Bundle bundle, BundleEvent event, List<ServiceRegistra
...
+        } else {
+            FormattingResultLog log = new FormattingResultLog();
+            log.warn("Expected bundles : {}, present bundles: {}", expectedBundles, registeredBundles);
```
Comment (raducotescu) [inline src/main/java/org/apache/sling/servlets/resolver/internal/bundle/BundledScriptTracker.java:606]:
```suggestion
            log.warn("Expected bundles : {}, registered bundles: {}", expectedBundles, registeredBundles);
```

**Review by raducotescu — COMMENTED** [review]
Apart from the tiny punctual comments, would it make sense to add support for multiple versions of the same bundle?

File: src/main/java/org/apache/sling/servlets/resolver/internal/bundle/BundledScriptTracker.java [hunk 6]
```diff
@@ -756,4 +805,16 @@ private Set<BundledRenderUnitCapability> reduce(Set<BundledRenderUnitCapability>
...
         return newSet;
     }
+
+
+    @ObjectClassDefinition
+    public @interface BundledScriptTrackerConfig {
+        
+        @AttributeDefinition(name="Mandatory Bundles", description="A list of symbolic bundle names the for which the "
+                + "script registration process must have been successfully completed for the health check to report ok.")
+        String[] mandatoryBundles();
```
Comment (raducotescu) [inline src/main/java/org/apache/sling/servlets/resolver/internal/bundle/BundledScriptTracker.java:815]:
```suggestion
        @AttributeDefinition(name="Mandatory Bundles", description="A list of symbolic bundle names for which the "
                + "script registration process must have been successfully completed for the health check to report ok.")
        String[] mandatoryBundles();
```

**Comment by raducotescu** [comment]
> @raducotescu wrote:
>> would it make sense to add support for multiple versions of the same bundle

> I don't know, what you mean with that in this context of this PR. Can you please elaborate?

Right now when we compare the expected to the registered bundles we only look at bundle symbolic names. Assuming we'd have two bundles B with different versions, say 1.0 and 2.0, and a configuration for an expected bundle B, we'd return ok from the health check if either of the two actual bundles B has registered its scripts.

**Comment by joerghoh** [comment]
@raducotescu ok, got it.

Assuming your case, wouldn't it be reasonable to assume that version 2.0 and version 1.0 have a large overlap of bundled scripts? That means: I don't know how the resolver would behave when both bundles would register a script for ``foo/bar/script``. In my opinion the result is unpredictable.

And for that I would suggest that we don't complicate this healthcheck until we support this.

File: src/main/java/org/apache/sling/servlets/resolver/internal/bundle/BundledScriptTracker.java (hunk 2 above)
Comment (joerghoh) [reply src/main/java/org/apache/sling/servlets/resolver/internal/bundle/BundledScriptTracker.java]: removed the "Sightly" from the property name.

**Pull request merged**